File: code/cli/sifter_cli/commands/records.py
from __future__ import annotations

import json
from typing import Optional

import typer
from sifter import Sifter
from sifter.errors import SifterError

from sifter_cli import config as cfg
from sifter_cli.output import auto_format, console, err_console

app = typer.Typer(help="Query and export records")


def _client(api_url: Optional[str], api_key: Optional[str], profile: str) -> Sifter:
    url, key = cfg.resolve(api_url, api_key, profile)
    return Sifter(api_url=url, api_key=key)


@app.command("list")
def list_records(
    sift_id: str = typer.Argument(...),
    limit: int = typer.Option(50, "--limit"),
    cursor: Optional[str] = typer.Option(None, "--cursor"),
    filter_json: Optional[str] = typer.Option(None, "--filter"),
    as_json: bool = typer.Option(False, "--json/--table"),
    api_url: Optional[str] = typer.Option(None, "--api-url"),
    api_key: Optional[str] = typer.Option(None, "--api-key"),
    profile: str = typer.Option("default", "--profile"),
) -> None:
    """List records from a sift."""
    client = _client(api_url, api_key, profile)
    try:
        filter_dict = json.loads(filter_json) if filter_json else None
        sift = client.sift(sift_id)
        if filter_dict or cursor:
            page = sift.find(filter=filter_dict, limit=limit, cursor=cursor)
            data = page.records
        else:
            data = sift.records(limit=limit)
        rows = [r.get("extracted_data", r) if isinstance(r, dict) else r for r in data]
        auto_format(rows, as_json)
    except SifterError as e:
        err_console.print(f"[red]Error:[/red] {e}")
        raise typer.Exit(2)


@app.command("query")
def query_records(
    sift_id: str = typer.Argument(...),
    question: str = typer.Argument(...),
    as_json: bool = typer.Option(False, "--json/--table"),
    api_url: Optional[str] = typer.Option(None, "--api-url"),
    api_key: Optional[str] = typer.Option(None, "--api-key"),
    profile: str = typer.Option("default", "--profile"),
) -> None:
    """Run a natural-language query against a sift."""
    client = _client(api_url, api_key, profile)
    try:
        sift = client.sift(sift_id)
        result = sift.query(question)
        results = result.get("results") or []
        auto_format(results, as_json)
    except SifterError as e:
        err_console.print(f"[red]Error:[/red] {e}")
        raise typer.Exit(2)


@app.command("export")
def export_records(
    sift_id: str = typer.Argument(...),
    output: str = typer.Option("records.csv", "--output", "-o"),
    api_url: Optional[str] = typer.Option(None, "--api-url"),
    api_key: Optional[str] = typer.Option(None, "--api-key"),
    profile: str = typer.Option("default", "--profile"),
) -> None:
    """Export records as CSV."""
    client = _client(api_url, api_key, profile)
    try:
        sift = client.sift(sift_id)
        csv_data = sift.export_csv()
        with open(output, "w", encoding="utf-8") as f:
            f.write(csv_data)
        console.print(f"[green]Exported to {output}[/green]")
    except SifterError as e:
        err_console.print(f"[red]Error:[/red] {e}")
        raise typer.Exit(2)

File: code/cli/src/commands/records.ts
import { writeFile } from "node:fs/promises";
import { Command } from "commander";
import { autoFormat, ok, err } from "../output.js";
import { makeClient, type GlobalOpts } from "../client.js";

export function recordsCommand(globals: () => GlobalOpts): Command {
  const cmd = new Command("records").description("Query and export records");

  cmd
    .command("list <sift-id>")
    .description("List records from a sift")
    .option("--limit <n>", "Max results", "50")
    .option("--cursor <cursor>", "Pagination cursor")
    .option("--filter <json>", "Filter as JSON e.g. '{\"total\":{\"$gt\":100}}'")
    .option("--json", "Output as JSON")
    .action(async (siftId, opts) => {
      try {
        const sift = await makeClient(globals()).getSift(siftId);
        const limit = parseInt(opts.limit);
        let rows: unknown[];

        const filter = opts.filter ? JSON.parse(opts.filter) as Record<string, unknown> : undefined;
        const page = await sift.find({ filter, limit, cursor: opts.cursor });
        rows = page.records;

        const data = rows.map(r => {
          const rec = r as Record<string, unknown>;
          return (rec["extracted_data"] as Record<string, unknown>) ?? rec;
        });
        autoFormat(data, !!opts.json);
      } catch (e) { err(String(e)); process.exit(2); }
    });

  cmd
    .command("query <sift-id> <question>")
    .description("Run a natural-language query against a sift")
    .option("--json", "Output as JSON")
    .action(async (siftId, question, opts) => {
      try {
        const sift = await makeClient(globals()).getSift(siftId);
        const { results } = await sift.query(question);
        autoFormat(results ?? [], !!opts.json);
      } catch (e) { err(String(e)); process.exit(2); }
    });

  cmd
    .command("export <sift-id>")
    .description("Export records as CSV")
    .option("-o, --output <file>", "Output file", "records.csv")
    .action(async (siftId, opts) => {
      try {
        const sift = await makeClient(globals()).getSift(siftId);
        const csv = await sift.exportCsv();
        await writeFile(opts.output, csv, "utf-8");
        ok(`Exported to ${opts.output}`);
      } catch (e) { err(String(e)); process.exit(2); }
    });

  return cmd;
}
